tests/numerics/diagonal_matrix_test.C

```diff
@@ -0,0 +1,222 @@
+// libmesh includes
+#include <libmesh/diagonal_matrix.h>
+#include <libmesh/parallel.h>
+#include <libmesh/auto_ptr.h>
+#include <libmesh/dense_matrix.h>
+#include <libmesh/id_types.h>
+#include <libmesh/numeric_vector.h>
+
+#include <libmesh/petsc_vector.h>
+#include <petsc.h>
+
+#include "libmesh_cppunit.h"
+#include "test_comm.h"
+
+#include <memory>
+#include <numeric>
+
+#define UNUSED 0
+
+using namespace libMesh;
+
+namespace
+{
+template <typename T>
+T
+termial(T n)
+{
+  if (n == 1)
+    return 1;
+  return n + termial(n - 1);
+}
+}
+
+class DiagonalMatrixTest : public CppUnit::TestCase
+{
+public:
+  void setUp()
+  {
+    _comm = TestCommWorld;
+    _matrix = libmesh_make_unique<DiagonalMatrix<Real>>(*_comm);
```

Comment:
Maybe try clearing this in the tearDown?  Perhaps CppUnit isn't destroying the DiagonalMatrixTest object until after the LibMeshInit destruction, for some reason, so then the tests pass but the destructor throws some kind of error?